interp/src.go

```diff
@@ -188,13 +190,87 @@ func pkgDir(goPath string, root, path string) (string, string, error) {
 		return "", "", fmt.Errorf("unable to find source related to: %q", path)
 	}
 
-	return pkgDir(goPath, previousRoot(root), path)
+	rootPath := filepath.Join(goPath, "src", root)
+	prevRoot, err := previousRoot(rootPath, root)
+	if err != nil {
+		return "", "", err
+	}
+
+	return pkgDir(goPath, prevRoot, path)
 }
 
 // Find the previous source root (vendor > vendor > ... > GOPATH).
-func previousRoot(root string) string {
-	splitRoot := strings.Split(root, string(filepath.Separator))
+func previousRoot(rootPath, root string) (string, error) {
+	// Arbitrary limit to prevent unforeseen infinite recursion.
+	maxDepth := 100
+	depthCounter := 0
+
+	const vendor = "vendor"
+
+	// TODO(mpl): maybe it works for the special case main, but can't be bothered for now.
+	parent, final := filepath.Split(rootPath)
+	if root != mainID && final != vendor {
+		// look for the closest vendor in one of our direct ancestors, as it takes priority.
+		prefix := path.Clean(strings.TrimSuffix(rootPath, root))
+		var vendored string
+		for {
```

Comment:
Here, you do not need to open the directories at each level and read/parse potentially all of their content. You could just construct the next path (up to `/`) and probe if the `path/vendor` is a valid directory using `os.Lstat` and `IsDir` method. That would simplify a lot (no defer, no closure, ...) and be always faster.

I do not see the need of `depthCounter` as well, as the number of paths up to `/` is always finite, and we are not walking through symlinks etc.

Reply:
> Here, you do not need to open the directories at each level and read/parse potentially all of their content. You could just construct the next path (up to `/`) and probe if the `path/vendor` is a valid directory using `os.Lstat` and `IsDir` method. That would simplify a lot (no defer, no closure, ...) and be always faster.

Agreed.

> I do not see the need of `depthCounter` as well, as the number of paths up to `/` is always finite, and we are not walking through symlinks etc.

Probably, yes. As this needed to be finished today (I won't be back until thursday, and they needed this fixed very soon) I opted for being overly safe just in case, and not spend too much time thinking about it.

Reply:
All done